internal/util/search/search.go

```diff
@@ -0,0 +1,178 @@
+	inout := &kio.LocalPackageReadWriter{
+		PackagePath:     resourcesPath,
+		NoDeleteFiles:   true,
+		PackageFileName: kptfile.KptFileName,
+	}
+	sr.Match = make(map[string][]string)
+	return kio.Pipeline{
+		Inputs:  []kio.Reader{inout},
+		Filters: []kio.Filter{kio.FilterAll(sr)},
+		Outputs: []kio.Writer{inout},
+	}.Execute()
+}
+
+// Filter parses input node and performs search and replace operation on the node
+func (sr *SearchReplace) Filter(object *yaml.RNode) (*yaml.RNode, error) {
+	filePath, _, err := kioutil.GetFileAnnotations(object)
+	if err != nil {
+		return object, err
+	}
+	sr.filePath = filePath
+
+	if sr.shouldPutLiteralByPath() {
+		return object, sr.putLiteral(object)
+	}
+
+	// traverse the node to perform search/put operation
+	err = accept(sr, object)
+	return object, err
+}
+
+// visitMapping parses mapping node
+func (sr *SearchReplace) visitMapping(object *yaml.RNode, path string) error {
+	return nil
+}
+
+// visitSequence parses sequence node
+func (sr *SearchReplace) visitSequence(object *yaml.RNode, path string) error {
+	// TODO: pmarupaka support sequence nodes
+	return nil
+}
+
+// visitScalar parses scalar node
+func (sr *SearchReplace) visitScalar(object *yaml.RNode, path string) error {
+	regexMatch, err := sr.regexMatch(object.Document().Value)
+	if err != nil {
+		return err
+	}
+
+	pathMatch, err := sr.pathMatch(path)
+	if err != nil {
+		return err
+	}
+
+	valueMatch := object.Document().Value == sr.Value || regexMatch
+
+	// at least one of path or value must be matched
+	if (valueMatch && pathMatch) || (valueMatch && sr.Path == "") ||
+		(sr.Value == "" && sr.ValueRegex == "" && pathMatch) {
+		sr.Count++
+		if sr.PutLiteral != "" {
+			object.Document().Value = sr.PutLiteral
+		}
+		if sr.filePath != "" {
+			sr.Match[sr.filePath] = append(sr.Match[sr.filePath], fmt.Sprintf("%s: %s", strings.TrimPrefix(path, PathDelimiter), object.Document().Value))
+		}
+	}
+	return nil
+}
+
+// regexMatch checks if ValueRegex in SearchReplace struct matches with the input
+// value, returns error if any
+func (sr *SearchReplace) regexMatch(value string) (bool, error) {
+	if sr.ValueRegex == "" {
+		return false, nil
+	}
+	re, err := regexp.Compile(sr.ValueRegex)
+	if err != nil {
+		return false, errors.Wrap(err)
+	}
+	return re.Match([]byte(value)), nil
+}
+
+// pathMatch checks if the traversed yaml path matches with the user input path
+// checks if user input path is valid
+func (sr *SearchReplace) pathMatch(yamlPath string) (bool, error) {
+	if sr.Path == "" {
+		return false, nil
+	}
+	// TODO: pmarupaka Path expressions should be supported
+	if !isAbsPath(sr.Path) {
+		return false, errors.Errorf(`invalid input path, must follow pattern e.g. foo.bar.baz`)
+	}
+	return sr.Path == strings.TrimPrefix(yamlPath, PathDelimiter), nil
+}
+
+// isAbsPath checks if input path is absolute and not a path expression
+// only supported path format is e.g. foo.bar.baz
+func isAbsPath(path string) bool {
```

Comment:
I suggest create a separate file to encapsulate path expression logic as it will get more complicated in the future. It will also need its own unit testing.

Reply:
Will do it as part of implementing path expressions story.